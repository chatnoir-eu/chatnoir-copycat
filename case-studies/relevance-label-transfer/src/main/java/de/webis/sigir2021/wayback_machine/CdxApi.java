package de.webis.sigir2021.wayback_machine;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.temporal.ChronoUnit;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.StringUtils;
import org.archive.format.warc.WARCConstants;
import org.archive.format.warc.WARCConstants.WARCRecordType;
import org.archive.io.warc.WARCRecordInfo;
import org.archive.uid.UUIDGenerator;
import org.archive.util.anvl.ANVLRecord;

import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class CdxApi {
	public static final RetryPolicy<String> RETRY_POLICY = new RetryPolicy<String>()
			  .handle(Exception.class)
			  .withBackoff(10, 240, ChronoUnit.SECONDS)
			  .withMaxRetries(3);
	
	public static WARCRecordInfo warcRecordForUrlBetween2009And2013(String id, String url) {
		try {
			return warcRecordForUrlBetween2009And2013ThrowsException(id, url);
		} catch(Exception e) {
			return null;
		}
	}
	
	private static WARCRecordInfo warcRecordForUrlBetween2009And2013ThrowsException(String id, String url) {
		String content = snapshotsForUrlBetween2009And2013(url);
		
		WARCRecordInfo ret = new WARCRecordInfo();
		ret.setType(WARCRecordType.response);
		ret.setRecordId(new UUIDGenerator().getRecordID());
		ANVLRecord extraHeaders = new ANVLRecord();
		extraHeaders.addLabelValue(WARCConstants.HEADER_KEY_URI, cdxRequest(url, "2009", "2013").toString());
		extraHeaders.addLabelValue("trec_target_id", id);
		extraHeaders.addLabelValue("trec_target_uri", url);
		ret.setExtraHeaders(extraHeaders);
		InputStream contentStream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
		ret.setContentLength(content.getBytes(StandardCharsets.UTF_8).length);
		ret.setMimetype("text/plain;charset=UTF-8");
		
		ret.setContentStream(contentStream);
		
		return ret;
	}
	
	public static String snapshotsForUrlBetween2009And2013(String url) {
		URL requestURL = cdxRequest(url, "2009", "2013");

		return snapshotsForUrlFailsave(requestURL);
	}

	private static String snapshotsForUrlFailsave(URL waybackUrl) {
		return snapshotsForUrlFailsave(waybackUrl, RETRY_POLICY);
	}
	
	private static String snapshotsForUrlFailsave(URL waybackUrl, RetryPolicy<String> retryPolicy) {
		return Failsafe.with(retryPolicy).get(() -> snapshotsForUrl(waybackUrl));
	}
	
	@SneakyThrows
	private static String snapshotsForUrl(URL waybackUrl) {
		return IOUtils.toString(waybackUrl, StandardCharsets.UTF_8);
	}
	
	@SneakyThrows
	private static URL cdxRequest(String url, String from, String to) {
		if(url.startsWith("https://")) {
			url = StringUtils.replaceFirst(url, "https://", "");
		}
		if(url.startsWith("http://")) {
			url = StringUtils.replaceFirst(url, "http://", "");
		}
		
		url = URLEncoder.encode(url, StandardCharsets.UTF_8.toString());
		
		return new URL("http://web.archive.org/cdx/search/cdx?url=" + url + "&from=" + from + "&to=" + to);
	}
}
