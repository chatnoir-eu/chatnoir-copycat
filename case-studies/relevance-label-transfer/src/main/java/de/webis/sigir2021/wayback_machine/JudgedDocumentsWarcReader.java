package de.webis.sigir2021.wayback_machine;

import java.io.ByteArrayOutputStream;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;
import java.util.stream.Collectors;

import org.apache.commons.io.IOUtils;
import org.archive.io.ArchiveRecord;
import org.archive.io.warc.WARCReader;
import org.archive.io.warc.WARCReaderFactory;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;

public class JudgedDocumentsWarcReader {

	private final Path dir;
	
	public JudgedDocumentsWarcReader(String dir) {
		this.dir = Paths.get(dir);
	}

	@SneakyThrows
	public String entries(int topic, String documentId) {
		Path warcFile = uniqueWARCFile(topic);
		WARCReader reader = WARCReaderFactory.get(warcFile.toFile());
		
		return recordWithId(reader, documentId);
	}

	private String recordWithId(WARCReader reader, String documentId) {
		for(ArchiveRecord record: reader) {
			if(documentId.equalsIgnoreCase((String) record.getHeader().getHeaderFields().get("trec_target_id"))) {
				return content(record);
			}
		}
		
		return null;
	}
	
	@SneakyThrows
	private static String content(ArchiveRecord record) {
		ByteArrayOutputStream ret = new ByteArrayOutputStream();
		record.dump(ret);
		
		return ret.toString(StandardCharsets.UTF_8.toString());
	}

	private Path uniqueWARCFile(int topic) {
		Path parentDir = dir.resolve("topic-" + topic);
		String[] entries = parentDir.toFile().list();
		
		if(entries == null || entries.length != 1) {
			throw new RuntimeException("Invalid dir: " + parentDir);
		}
		
		return parentDir.resolve(entries[0]);
	}

	@SneakyThrows
	public static Date parseDate(String line) {
		try {
			String ret = line.split("\\s+")[1];
			
			return sdf().parse(ret);
		} catch(Exception e) {
			throw new RuntimeException("'" + line + "'");
		}
	}
	
	public static String waybackUrl(String cdxRecord) {
		String date = cdxRecord.split("\\s+")[1];
		
		return "http://web.archive.org/web/" + date + "/" + targetUrl(cdxRecord);
	}
	
	public static String targetUrl(String cdxRecord) {
		return cdxRecord.split("\\s+")[2];
	}
	
	@SneakyThrows
	public static String resolveRedirect(String cdxRecord) {
		if(cdxRecord == null || parseResponseCode(cdxRecord) < 300 || parseResponseCode(cdxRecord)> 400) {
			return null;
		}
		
		return resolveRedirect(new URL(waybackUrl(cdxRecord)), 0);
	}
	
	@SneakyThrows
	public static String resolveRedirect(URL url, int depth) {
		if(depth > 4) {
			return url == null ? null : url.toString();
		}
		
		String html = htmlFailsafe(url.toString());
		String ret = extractRedirectMessageFromHtml(html);
		
		if (ret == null) {
			return url == null ? null : url.toString();
		} else {
			return resolveRedirect(new URL(ret), depth +1);
		}
	}
	
	@SneakyThrows
	private static String htmlFailsafe(String urlStr) {
		URL url = new URL(urlStr);
		return Failsafe.with(CdxApi.RETRY_POLICY).get(() -> html(url));
	}
	
	@SneakyThrows
	private static String html(URL url) {
		return IOUtils.toString(url);
	}

	private static SimpleDateFormat sdf() {
		SimpleDateFormat ret = new SimpleDateFormat("yyyyMMddHHmmss");
		ret.setTimeZone(TimeZone.getTimeZone("Europe/Berlin"));
		
		return ret;
	}
	
	private static Date centerOfCw12Crawl() {
		return parseDate("a 20120401000000 a");
	}

	public static int parseResponseCode(String line) {
		try {
			return Integer.parseInt(line.split("\\s+")[4]);
		} catch(Exception e) {
			return 0;
		}
	}

	public static List<String> keepResponsesInCw12CrawlingTime(List<String> entries) {
		return entries.stream()
				.filter(i -> responseInCw12CrawlingTime(i))
				.collect(Collectors.toList());
	}

	public static boolean responseInCw12CrawlingTime(Date date) {
		return parseDate("a 20111100000000 b").getTime() <= date.getTime() && 
				parseDate("a 20120800000000 b").getTime() >= date.getTime();
	}
	
	private static boolean responseInCw12CrawlingTime(String l) {
		if(l == null || l.trim().isEmpty()) {
			return false;
		}
		
		return responseInCw12CrawlingTime(parseDate(l));
	}

	private static Long diffToCW12(String line) {
		return Math.abs(centerOfCw12Crawl().getTime() - parseDate(line).getTime());
	}

	public String bestMatchingCW12SnapshotOrNull(int topic, String documentId) {
		return bestMatchingCW12SnapshotOrNull(entries(topic, documentId));
	}
	
	public String bestMatchingCW12SnapshotOrNull(String entryString) {
		if(entryString == null) {
			return null;
		}
		
		List<String> entries = Arrays.asList(entryString.split("\n"));
		if(entries.size() == 0) {
			return null;
		}
		
		entries = keepResponsesInCw12CrawlingTime(entries);
		entries = new ArrayList<>(entries.stream().filter(i -> 200 == parseResponseCode(i)).collect(Collectors.toList()));
		if(entries.size() == 0) {
			return null;
		}
		
		Collections.sort(entries, (a,b) -> diffToCW12(a).compareTo(diffToCW12(b)));
		return entries.get(0);
	}

	public String bestMatchingCW12RedirectOrNull(int topic, String documentId) {
		String entryString = entries(topic, documentId);
		if(entryString == null || bestMatchingCW12SnapshotOrNull(entryString) != null) {
			return null;
		}		

		List<String> entries = Arrays.asList(entryString.split("\n"));
		if(entries.size() == 0) {
			return null;
		}
		
		entries = keepResponsesInCw12CrawlingTime(entries);
		entries = new ArrayList<>(entries.stream().filter(i -> parseResponseCode(i) >= 300 && parseResponseCode(i) < 400).collect(Collectors.toList()));
		if(entries.size() == 0) {
			return null;
		}
		
		Collections.sort(entries, (a,b) -> diffToCW12(a).compareTo(diffToCW12(b)));
		return entries.get(0);
	}

	public static String extractRedirectMessageFromHtml(String html) {
		if(html == null) {
			return null;
		}
		
		Document doc = Jsoup.parse(html);
		Elements anchors = doc.select("p[class=impatient]>a[href]");
		
		if(anchors == null || anchors.size() != 1) {
			return null;
		}
		
		return anchors.get(0).attr("href");
	}
}
