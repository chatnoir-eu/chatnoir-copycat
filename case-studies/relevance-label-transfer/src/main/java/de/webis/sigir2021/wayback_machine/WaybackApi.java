package de.webis.sigir2021.wayback_machine;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.time.temporal.ChronoUnit;
import java.util.LinkedHashMap;
import java.util.Map;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.archive.format.warc.WARCConstants;
import org.archive.format.warc.WARCConstants.WARCRecordType;
import org.archive.io.warc.WARCRecordInfo;
import org.archive.uid.UUIDGenerator;
import org.archive.util.anvl.ANVLRecord;

import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class WaybackApi {

	public static final RetryPolicy<Pair<ByteArrayOutputStream, Map<String, String>>> RETRY_FINAL = new RetryPolicy<Pair<ByteArrayOutputStream, Map<String, String>>>()
			.handle(Exception.class)
			.withBackoff(3, 100, ChronoUnit.SECONDS).withMaxRetries(4);
	
	
	@SneakyThrows
	public static WARCRecordInfo warcRecord(String id, String cdxLine) {
		Pair<ByteArrayOutputStream, Map<String, String>> response = contentFailsave(cdxLine);
		Map<String, String> headers = response.getRight();
		WARCRecordInfo ret = new WARCRecordInfo();
		ret.setType(WARCRecordType.response);
		ret.setRecordId(new UUIDGenerator().getRecordID());
		InputStream contentStream = new ByteArrayInputStream(response.getLeft().toByteArray());
		
		ret.setContentLength(Integer.parseInt(headers.get("Content-Length")));
		ret.setMimetype(headers.get("Content-Type"));
		
		ANVLRecord extraHeaders = new ANVLRecord();
		extraHeaders.addLabelValue(WARCConstants.HEADER_KEY_URI, JudgedDocumentsWarcReader.waybackUrl(cdxLine));
		extraHeaders.addLabelValue("trec_target_id", id);
		extraHeaders.addLabelValue("trec_target_uri", JudgedDocumentsWarcReader.targetUrl(cdxLine));
		for(String header: headers.keySet()) {
			if(header == null || header.equalsIgnoreCase("Content-Length") || header.equalsIgnoreCase("Content-Type")) {
				continue;
			}
			
			extraHeaders.addLabelValue(header, headers.get(header));
		}
		
		ret.setExtraHeaders(extraHeaders);
		
		ret.setContentStream(contentStream);
		
		return ret;
	}
	
	@SneakyThrows
	public static WARCRecordInfo warcRecord(String id, URL url) {
		Pair<ByteArrayOutputStream, Map<String, String>> response = contentFailsave(url);
		Map<String, String> headers = response.getRight();
		WARCRecordInfo ret = new WARCRecordInfo();
		ret.setType(WARCRecordType.response);
		ret.setRecordId(new UUIDGenerator().getRecordID());
		InputStream contentStream = new ByteArrayInputStream(response.getLeft().toByteArray());
		
		ret.setContentLength(Integer.parseInt(headers.get("Content-Length")));
		ret.setMimetype(headers.get("Content-Type"));
		
		ANVLRecord extraHeaders = new ANVLRecord();
		extraHeaders.addLabelValue(WARCConstants.HEADER_KEY_URI, url.toString());
		extraHeaders.addLabelValue("trec_target_id", id);
		extraHeaders.addLabelValue("trec_target_uri", url.toString());
		for(String header: headers.keySet()) {
			if(header == null || header.equalsIgnoreCase("Content-Length") || header.equalsIgnoreCase("Content-Type")) {
				continue;
			}
			
			extraHeaders.addLabelValue(header, headers.get(header));
		}
		
		ret.setExtraHeaders(extraHeaders);
		
		ret.setContentStream(contentStream);
		
		return ret;
	}
	
	@SneakyThrows
	private static Pair<ByteArrayOutputStream, Map<String, String>> contentFailsave(String cdxLine) {
		URL waybackUrl = new URL(JudgedDocumentsWarcReader.waybackUrl(cdxLine));
		return contentFailsave(waybackUrl);
	}
	
	private static Pair<ByteArrayOutputStream, Map<String, String>> contentFailsave(URL waybackUrl) {
		return Failsafe.with(RETRY_FINAL).get(() -> content(waybackUrl));
	}
	
	@SneakyThrows
	private static Pair<ByteArrayOutputStream, Map<String, String>> content(URL waybackUrl) {
		
		URLConnection conn = waybackUrl.openConnection();
		
		ByteArrayOutputStream ret = new ByteArrayOutputStream();
		IOUtils.copy(conn.getInputStream(), ret);
		
		Map<String, String> headers = new LinkedHashMap<>();
		for(String header: conn.getHeaderFields().keySet()) {
			headers.put(header, conn.getHeaderField(header));
		}
		
		return Pair.of(ret, headers);
	}
}
