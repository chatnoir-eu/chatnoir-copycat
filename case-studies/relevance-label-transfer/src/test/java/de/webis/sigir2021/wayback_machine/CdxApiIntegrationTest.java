package de.webis.sigir2021.wayback_machine;

import org.approvaltests.Approvals;
import org.junit.Test;

import de.webis.sigir2021.wayback_machine.CdxApi;

public class CdxApiIntegrationTest {
	@Test
	public void approveWaybackSnapshotsForUrlWebis() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("http://webis.de");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveWaybackSnapshotsForUrlWebisWithHttpsProtocoll() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("https://webis.de");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveWaybackSnapshotsForUrlWebisWithHttpsProtocollAndTrailingBackslash() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("https://webis.de/");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveDogBreedsWikipediaArticle() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("https://en.wikipedia.org/wiki/Dog_breeds");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveDocumentFromWeb2009Track() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("http://www.cellphonenews.com/category/usa_canada_phones/3g_cell_phones");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveDocumentFromWeb2011Track() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("http://www.tottevents.com/index.php/Virtual-Reality/Laser-Skeet-Shooting");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveDocumentWithWhitespace() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("http://en.wikipedia.org/wiki/Biloxi, MS");
		
		Approvals.verify(actual);
	}
	
	@Test
	public void approveDocumentWithWhitespace2() {
		String actual = CdxApi.snapshotsForUrlBetween2009And2013("http://en.wikipedia.org/wiki/Biloxi, Mississippi");
		
		Approvals.verify(actual);
	}
}
