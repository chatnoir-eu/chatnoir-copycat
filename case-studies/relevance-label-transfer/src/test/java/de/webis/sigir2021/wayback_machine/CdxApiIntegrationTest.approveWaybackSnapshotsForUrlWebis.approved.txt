de,webis)/ 20110915232623 http://www.webis.de/ text/html 200 77OUBNP6JJZBM6RUFD6TIIGXWWOECMIB 601
de,webis)/ 20111110154922 http://www.webis.de/ text/html 200 6A7HTXBK7XME2IAROKAHLW7TAE4ACS2K 631
de,webis)/ 20130721031637 http://webis.de/ text/html 200 6A7HTXBK7XME2IAROKAHLW7TAE4ACS2K 629
